### ISSUE TITLE
Generating background from short segments

### ISSUE THREAD

**Issue by turpaultn** [justinsalamon/scaper#70]
I have a lot of different files for my background.
However, If I have a background files which is shorter than the duration specified for a sounscape, the end of my file has no background noise. Am I doing something wrong ?

If not, would it be possible to imagine one of this solution:
* 1) duplicate the background until the duration needed 
* 2) pick another background file and concatenate it with the previous one.

1) has the advantage of keeping a consistent environment in the background.

**Comment by justinsalamon** [comment]
@turpaultn if your background file is shorter than your soundscape duration then the current implementation should concatenate the bg file to itself as many times as needed to fill the soundscape duration (i.e. option 1 you describe): https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1651-L1685

Are you not observing this behavior? If so, can you please post a MWE here so I can try and reproduce this issue?

**Comment by turpaultn** [comment]
```python
from scaper import Scaper

sc = Scaper(10, "synthetic/audio/eval/soundbank/foreground", "synthetic/audio/eval/soundbank/background")
sc.add_background(("const","freesound"), ("const","synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav"), ("const", 0))
fname = "test_bg"
sc.generate(f"{fname}.wav", f"{fname}.jams")


sc = Scaper(10, "synthetic/audio/eval/soundbank/foreground", "synthetic/audio/eval/soundbank/background")
sc.add_background(("const","freesound"), ("const","synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav"), ("const", 0))
fname = "test_bg_fg"
sc.add_event(("const", "Blender"), ("const", "synthetic/audio/eval/soundbank/foreground/Blender/56563_0.wav"),
             ("const", 0), ("const", 5), ("const", 5), ("const", 6), ("const", 0), ("const", 1))
sc.generate(f"{fname}.wav", f"{fname}.jams")
```
I did a minimum tar archive here (link available 2 weeks):  [audios.tar.gz](https://filesender.renater.fr/?s=download&token=65952e9d-ce88-88af-cc2b-8546ad2b54a5)
Otherwise, the data are part of : [DESED_synth_soundbank.tar.gz](https://zenodo.org/record/3567927/files/DESED_synth_soundbank.tar.gz?download=1)


The behavior observed:
The background file is 3 seconds, the duration wanted is 10 seconds.
- In the first case with no foreground events, the audio file lasts 3 seconds (same time as the background)
- In the second case, the audio lasts 10 seconds and the background stops after 3 seconds.

**Comment by turpaultn** [comment]
After a little debug part, it seems the problem comes from the trimming part.
https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1670-L1673
Indeed, in the actual code, we trim to the length of the background event.
However, the value of background `event_duration` is trimmed if it is too long, but nothing is done if it is too short.
https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1303-L1313

I don't know how much it could impact the code to change this:
```python
 # Then trim 
 cmb.trim(e.value['source_time'], 
          e.value['source_time'] + 
          self.duration) 
```

Otherwise, I tried on my side to add a condition L1331-1355 when (event_duration < self.duration), however, I have a problem with the condition checked here `if source_time + event_duration > source_duration`:
https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1370-L1376

Hope this can help.

**Comment by pseeth** [comment]
Hm, so just looking at this briefly - isn't the background concatenated here in core.py:

https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1683-L1685

That's a bit further down. If that line is working as expected, perhaps the error is here in ntiles.

https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1661-L1662

I'll reproduce using your code first and get back to you.

**Comment by pseeth** [comment]
Okay ran your example. I can reproduce the bug. So looking at what sox actually does under the hood, it seems the command it executes is as follows:

```
sox -D -V2 --combine concatenate -v 1 synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav -v 1 synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav -v 1 synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav -v 1 synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav -c 1 /var/folders/5h/6vxlzhys259350xz4nb3xbym0000gn/T/tmp3gck8ybd.wav rate -h 44100.000000 trim 0.000000 2.912187
```

for your example. ntiles is set correctly (2.91 * 4 > 10). This appears to not apply the trim to each audio file and then combine, but rather first concatenates and then trims it to event_duration, which is the behavior you're seeing. The correct sequence of commands should be like:

```
sox -D -V2 --combine concatenate -v 1 synthetic/audio/eval/soundbank/background/freesound/noise-free-sound-0145.wav -c 1 temp.wav rate -h 44100.000000 trim 0.000000 2.912187
sox -D -V2 --combine concatenate -v 1 temp.wav -v 1 temp.wav -v 1 temp.wav -v 1 temp.wav -c 1 out.wav rate -h 44100.000000 trim 0 10
```

Will dig into this a bit more.

**Comment by turpaultn** [comment]
Thank you very much.

**Comment by pseeth** [comment]
Here's the entirety of the fix (backwards compatible). Turns out the event duration was being screwed up causing a very short trim. The event duration changes were introduced in #53 (by me - oops). But in the test cases, we never try anything with that short of background files. All of the background files are between 45 and 60 seconds or so.

https://github.com/pseeth/scaper/commit/2e3a9fe71fc4c40c7753a371db9fbcf91d5ecf74

This is based off of #55 which will be pulled in to master soon. Right after that, we can pull this branch into master to fix this issue, or pull it into #55 before merging #55  into master. But for now, you can just pull in the `max` change if you want to fix your own code.

**Comment by justinsalamon** [comment]
OK friends, I've given this a deep dive and here's what I've found:

* `event_duration` is forced to match `source_duration` when `source_duration` < `event_duration`, both for FG and BG events. This happens here: https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1303-L1313
* This means that say `sc.duration = 10`, but the duration of your source file for the BG is 3, then the `event_duration` for the BG will be modified to 3.
* The first place where this will cause a problem is in these lines where the event duration added to the annotation will be incorrect (assuming we self-concatenate): https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1532-L1535
* If I'm not mistaken, concatenation is the first thing the sox combiner does. This means that the trim operation in these lines will take the concatenated background and trim it back to `event_duration`, which has been changed back to `source_duration`, leading to the observed bug: https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1670-L1673
* For reference, these are the lines that do the concatenation (as well as all other combiner operations that follow): https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1682-L1685

I don't believe this issue was introduced by @pseeth, I think it was there all along and never got caught because none of our tests use background source files that are shorter than the specified soundscape duration, and that's on me! It's also another great example of why 100% coverage is not necessarily 100% coverage... :)

OK, now that the bug is isolated, it's time to get to a fix. First of all, we need to conceptually choose what should happen here, out of 2 options. If the background source file is shorter than the specified duration, we can:
1. self-concatenate it until it reaches the specified duration (as was originally intended though never properly implemented)
2. pad the background source file with silence to reach the soundscape duration.

Since there is no "correct" behavior here, it's up to us to determine what's the most useful behavior. If we wanted to be fancy, we could actually support both options with one being the default behavior and the other activated via a flag. I'm curious to hear @pseeth @turpaultn & @jtcramer 's opinion on what you'd consider the most convenient default behavior.

In terms of implementation, here's my 2c regarding each of the options:

**IMPLEMENTING CONCATENATION**
* It seems to me like the correct way to handle this would be to modify the code in https://github.com/justinsalamon/scaper/blob/7fb77a92cf9fda15aae7defb5a3d826f87877b4e/scaper/core.py#L1303-L1313 such that the event_duration only gets changed for foreground events. 
* This means that for background events, the event duration is kept unchanged (=soundscape duration) even if the background source is shorter than the soundscape duration
* This will result in the annotation having the correct duration for the BG (=soundscape duration)
* This will also mean the trim operation will function correctly on the background

So, if I'm not mistaken (and I may be so we definitely need tests for this), this simplest/cleanest solution will be to add an IF statement in the lines linked to above such that they are skipped if `is_background` is True.

**IMPLEMENTING SILENCE PADDING**
This one's a little trickier. Ignoring the annotation for a moment - we could do as proposed above, i.e. not modify the event duration for background events. Then, in `_generate_audio` we'd have to replace the tiling behavior with a zero padding behavior, but the rest, I believe, would remain unchanged and should work correctly (i.e. trimming should work fine).

The complication arises because does I'm not sure it makes sense for the annotation to say the background event lasts the full soundscape duration when in reality it's padded with silence. So the alternative would be to NOT modify lines L1303-L1313 such that if the BG source file is shorter, it modifies the event duration. Then we'd still pad it with silence, and modify the trim operation for background events to ignore the event duration and instead always use the soundscape duration.

Then the even more complex scenario is if we want to support both.

**SUPPORTING BOTH CONCAT AND PADDING**
Perhaps the way to achieve this is through some IF statements in both _instantiate_event and _generate_audio. Basically:
* IF bg-mode is CONCAT, then _instantiate_event will not modify event_duration (so it stays equal to the soundscape duration) and _generate_audio will apply concatenation via tiling.
* IF bg-mode is PAD, then _instantiate_event will modify event_duration to match the source_duration, which will annotate the BG as being shorter than the soundscape, and in _generate_audio we instead apply padding AND trim to the soundscape duration.


Personally, unless there's strong evidence that there's need for supporting the PAD variant, my preference would be to only fix the code so that the CONCAT approach works correctly. My motivation is that:
1. There are a lot of moving parts here and it will be easier to make sure we get it right if we limit the behavior to just one option. It will also be faster to implement and cleaner to test.
2. Supporting both modes could lead to confusion among users, especially if each mode leads to a different BG duration being stored in the JAMS annotation. 

Please chime in!

**Comment by pseeth** [comment]
Super helpful breakdown! I'm okay with only implementing concatenation - that's what the docs say will happen so I think that's what we should keep doing. I think that if the system is concatenating, a warning should be thrown to the user telling them that.

I agree that adding the if statement is the cleanest way to do it. To make the test case, I'll need to add a short background audio file to the repo. I think I can take the one posted in #70, which was shorter than the duration.

To test it, this might require a new regression test, right? So that would be added in `create_regression_data.py`, I believe. The tests should be on two levels - the first is that the audio corresponding to a soundscape with a short background event added to it has a duration matching what we expect. Second, the audio should contain the background concatenated with itself a few times. I believe the 2nd condition should be tested via some additional regression data.

**Comment by turpaultn** [comment]
Thank you very much Justin, very well explained !! It's easier to speak about it like this. 
I agree with @pseeth it would be better to do the concatenation since that's what is advertised for now, and nobody seemed to complain about it.

#### New scenario: choosing another file
Moreover, in a scenario when we do sound event detection, it could be interesting to:
- be able to pick another file if the background is not long enough (to avoid including something non stationary in the background that could lead to false detection)

So our two interesting scenarios are:
- bg-mode is CONCAT
- bg-mode is DROP then we pick another file that will match the right duration (warning could be raised)

#### Silence cases (that I'm using just for examples)
Maybe someone could explain to me a scenario where there is padding for the background file since I do not see any.

I feel like if you mix something with silence, isn't your events become a foreground events with silence background ? 
Because that's what I do in that case:
- Silence background + foreground events
- Remove the classes I did not want from my annotations when I gather them if I want them to be "background events".

Other scenario (opposite):
- background event (I want it to last the full file) + foreground events
- Get the annotations from background and foreground since I want to recognize both of them.

Sorry if I'm unclear.

**Comment by justinsalamon** [comment]
Thanks for your input @pseeth and @turpaultn !

OK, looks like we're in agreement about only supporting the concatenate scenario. @turpaultn also proposed a "drop" scenario, which I think we can definitely consider but I suggest we leave that for a separate PR to keep things clean.

Now that we're in agreement on the intended behavior and the correct fix, @pseeth wanna take a stab at this since you already have a PR open?

**Comment by pseeth** [comment]
I've implemented the fix discussed and added a test case. I wanted to document one funny bug I came across in implementing our fix here. So I added the fix that `if isbackground: event_duration = self.duration`. This resulted in an issue later on in `_instantiate_event` here:

https://github.com/justinsalamon/scaper/blob/b7aea91a467374945fabc21439dc1b4de374ae72/scaper/core.py#L1400-L1407

So the logic here is to find a satisfiable `source_time`. At line 1405, if the start time of the event plus the duration of the event is bigger than then source duration, then we "fix" the source time to be source duration - event duration. This works great because earlier we fix event duration to never exceed the duration of the source audio file. But when we make the event_duration fixed to self.duration if it's background, AND if the background is shorter than the duration of the soundscape, then we enter an infinite loop because `source_time = source_duration - event_duration < 0`! So the fix is simple:

```
source_time = max(0, source_duration - event_duration)
```

And everything works. PRing now but wanted to document.
